### PR TITLE
fix(core,hono): generalize the #3634 tag-identity fix and fix the same bug in hono

### LINKED ISSUE #3634: filterParams helper is missing for some tags (tags-split + Angular)
### Description

When generating an Angular service in tags-split mode, the generated service for certain tags calls the `filterParams(...)` helper but never defines or imports it, resulting in `TS2304: Cannot find name 'filterParams')`.

| Endpoint | Tag | Generated file | Result |
| --- | --- | --- | --- |
| `GET /widgets` | `Widget` | `widget/widget.service.ts` | defines `filterParams` |
| `GET /ab-widgets` | `AB Widget` | `ab-widget/ab-widget.service.ts` | calls `filterParams`, never defines it |

### Output client

angular

### Configuration (orval.config)

```ts
module.exports = {
  repro: {
    input: { target: './openapi.yaml' },
    output: {
      mode: 'tags-split',
      client: 'angular',
      target: './src/generated/services',
      schemas: './src/generated/model',
    },
  },
};
```

### Environment

Node 24.11.0
Orval 8.18.0
Using Ubuntu

### Expected behavior

Every generated service that calls `filterParams` should also define (or import) it, regardless of how the tag is spelled.

### Actual behavior

`TS2304: Cannot find name 'filterParams')` errors for some generated files.

### OpenAPI document (minimal, if applicable)

<details>
<summary>OpenAPI spec</summary>

```yaml
openapi: 3.0.0
info:
  title: Orval tags-split filterParams repro
  version: 1.0.0
paths:
  # Control: a single-word tag. Generates correctly.
  /widgets:
    get:
      tags: [Widget]
      operationId: listWidgets
      parameters:
        - name: search
          in: query
          schema:
            type: string
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                type: array
                items:
                  type: object

  # Trigger: a tag whose first word is an acronym ("AB") followed by a space.
  # Generates a broken service (calls filterParams but never defines it).
  /ab-widgets:
    get:
      tags: [AB Widget]
      operationId: listAbWidgets
      parameters:
        - name: search
          in: query
          schema:
            type: string
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                type: array
                items:
                  type: object

```

</details>

### Additional context

Minimal reproduction repo (generated with Claude Code): https://github.com/jeroentvb/orval-filterparams-repro

### PR DESCRIPTION
**Pull request by the-ult**
## Summary

Fixes #3634 — the `TS2304: Cannot find name 'filterParams'` bug for acronym / multi-word tags in Angular `tags-split` — and goes further:

1. **Generalizes** the root cause into a single source of truth in `@orval/core`, so the whole *class* of bug is fixed rather than the one Angular call site.
2. **Fixes the same-class bug in `@orval/hono`** — a separate, previously-unreported instance.

> **Credit.** @wadakatu independently arrived at the minimal Angular fix in #3640 (now closed in favor of this PR) — we converged on nearly identical Angular changes. Thanks to them for the catch. This PR keeps that fix and lifts it into a shared core helper that closes the bug class everywhere, plus the hono fix.

Closes #3634.

## Root cause

"Which file does an operation belong to in `tags` / `tags-split` mode?" is a real concept with no home. The rule — `kebab(operation.tags[0] ?? 'default')` — was hand-copied across **6 sites in 4 packages**, in two slightly different variants. Each copy independently decided how to *derive* the tag key and how to *compare* against it, so the build-the-key side and the look-up-the-key side could silently disagree. `@orval/angular` looked up with `camel()` while the file was keyed with `kebab()`:

| input | `kebab()` (file key) | `camel()` (old lookup) |
| --- | --- | --- |
| `"AB Widget"` | `ab-widget` | `aBWidget` |
| `"ab-widget"` | `ab-widget` | `abWidget` |

`aBWidget !== abWidget` → the operation became invisible to its own tag file, `filterParams` was suppressed, but the operation body still called it. Single-word tags were unaffected (`kebab` is a no-op).

## The generalization

A single source of truth in `@orval/core`, with every consumer routed through it:

- `getTagKey(tag?)` — canonical, idempotent (`kebab`-based) bucket key
- `getOperationTagKey(op)` — key for an operation's primary tag
- `isOperationInTagBucket(op, key)` — canonical-vs-canonical membership (use instead of hand-rolling `op.tags[0] === key`)

Migrated: the core `target-tags` writer, `@orval/angular`, `@orval/query`, `@orval/mock`, and `@orval/hono`. After this there are **no** remaining `kebab(...tags[0]...)` derivations or raw `tags[0] === …` comparisons anywhere — the bug class is closed at the source.

## Also fixes: `@orval/hono` (same class, previously unreported)

Found while auditing for the same pattern — no existing issue tracked it. In `tags-split` mode with a multi-word/acronym tag, hono was broken two ways:

1. `getVerbOptionGroupByTag` keyed by the **raw** `tags[0]`, so handler/zod/context files were written into a raw `AB Widget/` directory while the composite route imported them from the canonical `ab-widget/` directory → unresolved imports.
2. The composite-route handler filter compared a raw `tags[0]` against the canonical key and silently produced an **empty import block**, so the tag's handlers were never wired into the routes.

Both now go through the shared helper and emit consistent, canonical paths.

## Tests

- **Unit** (`@orval/core`, `@orval/angular`): key derivation, idempotency, and acronym / default-bucket matching.
- **CI-gated regression** (`test:snapshots` + `orval-tests build`): `tests/specifications/issue-3634.yaml` (`Widget` control + `AB Widget` trigger, both with query params), wired into an angular `tags-split` config **and** a hono composite-routes `tags-split` config. A regression fails both the snapshot diff and the TypeScript compile (`TS2304`).

  > The existing `Dup Tag`/`dup tag` duplicate-tag fixtures do **not** cover this — they all camelCase to the same `dupTag`, so the old (buggy) matching grouped them correctly. Only an acronym-prefixed tag exposes the camel-vs-kebab divergence.

Verified end-to-end with `AB Widget`: angular `ab-widget.service.ts` defines `filterParams` alongside its call sites, and hono `routes.ts` imports/wires the tag's handlers from a matching `ab-widget/` directory. **Single-word tags are unaffected — all existing sample snapshots are byte-identical.** Local `build`, `typecheck`, `lint`, `test`, `test:cli`, `test:snapshots` all green.

## Related

- **#3640** — @wadakatu's independent minimal Angular fix for #3634 (closed in favor of this); this PR generalizes the same root cause and adds the hono fix.
- Same family (closed): **#2998** ("filterParams never generated, angular + tags-split") and its partial fix **#3018**; earlier symptoms **#3103** / **#3104**.
- See also **#3512** — a `@orval/hono` `tags-split` path-correctness bug in the same area (already fixed separately).

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed tag normalization for multi-word and acronym tags (e.g., "AB Widget") across code generators, ensuring consistent kebab-case directory and filename generation.
  * Improved handling of untagged operations in tag-based code generation modes.

* **Tests**
  * Added test coverage for tag normalization edge cases and multi-word tag scenarios.

### CODE REVIEW

File: packages/core/src/utils/tags.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * that is already a canonical key. Other case functions (`camel`, `pascal`) are
+ * NOT safe here: they do not round-trip through the bucket key for tags
+ * containing acronyms or spaces (e.g. `"AB Widget"`), which is exactly the class
+ * of bug this module exists to prevent.
+ *
+ * Missing or empty tags map to the implicit {@link DefaultTag} bucket.
+ */
+export function getTagKey(tag?: string): string {
+  return kebab(tag ?? DefaultTag);
+}
```
[the remark this answers is not in the document]
Comment (the-ult) [reply packages/core/src/utils/tags.ts:26]: Accepted in
  6127e46a7. getTagKey now trims and falls back to the DefaultTag bucket for empty/whitespace tags, and
  isOperationInTagBucket guards on tagKey == null so only undefined is the no-filter case. JSDoc updated
  and unit tests added.

[the remark this answers is not in the document]
Comment (the-ult) [reply packages/core/src/utils/tags.ts:26]: Accepted in
  6127e46a7 — getTagKey now trims and falls back to the default bucket for empty/whitespace tags,
  matching the docstring. Unit tests added.

File: packages/core/src/utils/tags.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * absent `tagKey` matches every operation (the "no tag filter" case).
+ *
+ * Prefer this over hand-rolling `operation.tags[0] === tagKey`: a raw tag
+ * compared against a canonical key silently fails for multi-word/acronym tags.
+ */
+export function isOperationInTagBucket(
+  operation: { tags: string[] },
+  tagKey?: string,
+): boolean {
+  if (!tagKey) return true;
```
[the remark this answers is not in the document]
Comment (the-ult) [reply packages/core/src/utils/tags.ts]: Accepted in
  6127e46a7 — isOperationInTagBucket now guards on tagKey == null, so only undefined disables filtering;
  an empty string normalises to the default bucket. JSDoc and tests updated.

File: packages/core/src/utils/tags.test.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    ]) {
+      const once = getTagKey(tag);
+      expect(getTagKey(once)).toBe(once);
+    }
+  });
+
+  it('falls back to the default bucket for missing tags', () => {
+    expect(getTagKey()).toBe('default');
+    expect(getTagKey(undefined)).toBe('default');
+  });
```
[the remark this answers is not in the document]
Comment (the-ult) [reply packages/core/src/utils/tags.test.ts:38]: Accepted in
  6127e46a7 — added assertions covering empty/whitespace tags falling back to the default bucket, plus a
  companion test for isOperationInTagBucket.

File: packages/core/src/writers/target-tags.ts [hunk 4]
```diff
@@ -203,10 +208,7 @@ export function generateTargetForTags(
           // Operations can have multiple tags, but they are grouped by the first
           // tag, therefore we only want to handle the case where the tag
           // is the first in the list of tags.
-          .filter(
-            ({ tags }) =>
-              tags.map((tag) => kebab(tag)).indexOf(kebab(tag)) === 0,
-          )
+          .filter(({ tags }) => tags.length > 0 && getTagKey(tags[0]) === tag)
```
[the remark this answers is not in the document]
Comment (the-ult) [reply packages/core/src/writers/target-tags.ts]: Accepted in ab6e71fc4. The footer's `operationNames` now filters the default-normalised `operations` array (the one mapped through `addDefaultTagIfEmpty`) via the canonical `isOperationInTagBucket`, instead of raw `builder.operations` with a `tags.length > 0` guard. The implicit `default` bucket file therefore keeps its per-operation `*ClientResult` exports for untagged operations.

File: packages/angular/src/utils.ts [hunk 5]
```diff
@@ -156,15 +154,8 @@ export const getRelevantVerbOptionsForTag = (
   const allVerbOptions = Object.values(verbOptions);
   if (!tag) return allVerbOptions;
```
[the remark this answers is not in the document]
Comment (the-ult) [reply packages/angular/src/utils.ts]: Accepted in ab6e71fc4. `getRelevantVerbOptionsForTag` now guards on `tag == null` so only an absent tag means "no filter"; an empty/whitespace tag normalises to the `default` bucket via `isOperationInTagBucket`, matching the core writer. Added a regression test asserting an empty-string tag resolves to the default bucket rather than matching every operation.

**Review by melloware — APPROVED** [review]
